feature/CONTROLLER Update controller

ichabanivan opened the pull request:
- Added DEBUG Mode
- Added types for selector
- Added types for actions
- Added types for subscribe
- Update README
- Update reducer store data to next model { connected: boolean; actual: I; initial: I; }

- Remove named export for controller
- Add checking for all input data
- Add utils (isGeneratorFunction, isString, isBoolean, isArray, isObject)
- Simplify record for useEffect in useReducerSubscribe and useSubscribe
- Update documentation
- Add comments

- Create example
- Update README.md (Changed examples, Describe API)
- Update npm ignore
- Create todo.md
- Update tsconfig
- Create different actions (all,actual,initial,connected)

- Remove extra types in the controller
- Change logs to different view
- Create  common selector for CSD

ichabanivan committed:
feature/CONTROLLER-update-controller

- Implement DEBUG Mode

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update documentation with DEBUG Mode

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update package json scripts

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update default prefix for controller

ichabanivan committed:
feature/CONTROLLER-update-controller

- Added types for selector
- Added types for actions
- Added types for subscribe
- Update README
- Update reducer store data to next model { connected: boolean; actual: I; initial: I; }

ichabanivan committed:
feature/CONTROLLER-update-controller

- Remove named export for controller
- Add checking for all input data
- Add utils (isGeneratorFunction, isString, isBoolean, isArray, isObject)
- Simplify record for useEffect in useReducerSubscribe and useSubscribe
- Update documentation
- Add comments

ichabanivan committed:
feature/CONTROLLER-update-controller

- Create example
- Update README.md (Changed examples, Describe API)
- Update npm ignore
- Create todo.md
- Update tsconfig
- Create different actions (all,actual,initial,connected)

ichabanivan commented:
![image](https://user-images.githubusercontent.com/16426586/99917891-0029b680-2d1c-11eb-958f-2b6471253572.png)

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update README.md

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update Example

ichabanivan committed:
feature/CONTROLLER-update-controller

- Remove extra types in the controller
- Change logs to different view
- Create  common selector for CSD

ichabanivan committed:
feature/CONTROLLER-update-controller

- Codestyle

ichabanivan committed:
feature/CONTROLLER-update-controller

- Remove default export Controller

ichabanivan committed:
feature/CONTROLLER-update-controller

- Codestyle

sajera commented on example/store.ts:
```diff
@@ -0,0 +1,23 @@
+  combineReducers({
+    // ...your other reducers here
+    // you have to pass controllerReducer under 'controller' key,
+    controller: controllerReducer,
```
as addition ability we may implement method to provide custom redux name for example
```javascript

...controllerReducerCustom('myCustomReducrName'); //  => { [name]:  controllerReducer }
```

sajera commented on src/controller.ts:
```diff
@@ -3,71 +3,108 @@
-    this.subscriber = subscriber;
-    // TODO forbid to pass properties
+    // NOTE Setup DEBUG value if present and is valid
+    this.DEBUG = isBoolean(DEBUG) ? DEBUG : false;
```
use local storage to enable debug mode

sajera commented on src/utils.ts:
```diff
@@ -15,15 +15,32 @@ const idCounter: Record<string, number> = {};
-  if (!idCounter[prefix]) {
-    idCounter[prefix] = 0;
+export const uniqueId = (prefix = 'controller_') => {
+  if (!counter[prefix]) {
```
should generate not only unique but new one each time

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update example
- Update README
- DefaultActions = 'UPDATE_CTRL' | 'CLEAR_CTRL'

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update todo list

ichabanivan committed:
feature/CONTROLLER-update-controller

- Update todo list
